Add zooming functionality

gelaechter opened the pull request:
Hey, this should allow scaling the app and maybe even be a quick fix for #124 and #130.
I noticed that we aren't able to use the plus key in hotkeys. There doesn't seem to be a way around it, since [Mantines hotkey parser splits hotkey strings at pluses](https://github.com/mantinedev/mantine/blob/8c12a76c56da51af34213f18dd67c8b72a0ddb44/src/mantine-hooks/src/use-hotkeys/parse-hotkey.ts#L18). Do you maybe know a way around this to allow mod+plus and mod+minus to be the hotkeys for zooming?

gelaechter committed:
Add zooming functionality

jeffvli commented:
Thanks for the PR!

> I noticed that we aren't able to use the plus key in hotkeys. There doesn't seem to be a way around it, since [Mantines hotkey parser splits hotkey strings at pluses](https://github.com/mantinedev/mantine/blob/8c12a76c56da51af34213f18dd67c8b72a0ddb44/src/mantine-hooks/src/use-hotkeys/parse-hotkey.ts#L18). Do you maybe know a way around this to allow mod+plus and mod+minus to be the hotkeys for zooming?

Apparently you can get around this by modifying the `debouncedSetHotkey` function in `hotkey-manager-settings.tsx` to replace any `+` keys with just `equal`.

```typescript
if (!IGNORED_KEYS.includes(e.key)) {
  if (e.code.includes('Numpad')) {
    if (e.key === '+') keys.push('numpadadd');
    else if (e.key === '-') keys.push('numpadsubtract');
    else if (e.key === '*') keys.push('numpadmultiply');
    else if (e.key === '/') keys.push('numpaddivide');
    else if (e.key === '.') keys.push('numpaddecimal');
    else keys.push(`numpad${e.key}`.toLowerCase());
  } else if (e.key === '+') {
    keys.push('equal');
  } else {
    keys.push(e.key?.toLowerCase());
  }
}
```

jeffvli commented on src/renderer/features/settings/components/general/application-settings.tsx:
```diff
@@ -53,6 +56,30 @@ export const ApplicationSettings = () => {
+            });
+            localSettings.setZoomFactor(e);
+          }}
+        />
```
This would be better as on `onBlur` since it's a bit weird when you modify it and it instantly updates with a 1 digit value. Also stray console.log is being output here.

jeffvli commented on src/renderer/layouts/default-layout.tsx:
```diff
@@ -52,7 +54,26 @@ export const DefaultLayout = ({ shell }: DefaultLayoutProps) => {
+        zoomFactor: newVal,
+      },
+    });
+    console.log(settings.zoomFactor);
```
Stray console.log

jeffvli commented on src/renderer/layouts/default-layout.tsx:
```diff
@@ -52,7 +54,26 @@ export const DefaultLayout = ({ shell }: DefaultLayoutProps) => {
+  };
+  localSettings?.setZoomFactor(settings.zoomFactor);
+
+  useHotkeys([[bindings.zoomIn.hotkey, () => updateZoom(5)]]);
+  useHotkeys([[bindings.zoomOut.hotkey, () => updateZoom(-5)]]);
   useHotkeys([[bindings.globalSearch.hotkey, () => handlers.open()]]);
```
You can combine this into a single useHotkeys hook instead of instantiating it three times. 
- [Example from docs](https://mantine.dev/hooks/use-hotkeys/)

Also, since this is a desktop-only feature you should probably conditionally add these hotkeys.

```typescript
const zoomHotkeys: HotkeyItem[] = [
  [bindings.zoomIn.hotkey, () => updateZoom(5)],
  [bindings.zoomOut.hotkey, () => updateZoom(-5)],
];

useHotkeys([
  [bindings.globalSearch.hotkey, () => handlers.open()],
  ...(isElectron() ? zoomHotkeys : []),
]);
```

gelaechter committed:
Change zoomFactor input behaviour
Make zoom hotkey binding conditional
Allow plus to be used in hotkeys

gelaechter commented:
Thanks for your review! I implemented your comments.
I also implemented your suggestion to replace + with 'equals'. It makes sense that this works on the US layouts since plus is basically shift+equals but keep in mind that this doesn't necessarily work on other layouts where plus is a dedicated key. At least it doesn't work on the German layout. But again, I don't believe there is something we can do about that.

jeffvli commented:
Have you tried using `add`?

According to [this](https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values) it's a numpad value but it might work still.

gelaechter commented:
Yep, I tried a few: add, Add, plus, Plus, but none work.

jeffvli commented:
Seems like it might not be possible then unless we switch to a different hotkeys handler or we fork the useHotkeys hook into this repo and modify it to work. Unless you want to try that first this should be good to merge.

gelaechter commented:
You can merge it now, I might open an issue at mantines repo later.
